File: src/utils/shotDetection.ts
// =============================================================================
// SHARED SHOT-DETECTION PRIMITIVES
// =============================================================================
// Single source of truth for shuttle-trajectory + pose-classification shot
// detection. Consumed by useShotSegments (auto-pause + shot speed list) and
// useAdvancedAnalytics (rally detection). Python mirror in
// backend/rally_detection.py — keep algorithms and threshold defaults aligned.
//
// Algorithmic parity notes:
//   * dot-threshold math matches backend Python `_detect_shots`:
//       if cosAngleMax < 0: threshold = cosAngleMax * sqrt(s1sq * s2sq)
//       else:              threshold = 0
//       reject when dot >= threshold  (i.e. trigger when dot < threshold)
//     Stable against zero-speed edge cases (no division by zero).
//   * minSpeedSq is compared against per-step displacement squared
//     (Δx² + Δy² between sampled points), the same metric backend and
//     useAdvancedAnalytics use. It is NOT a per-second velocity magnitude.
//   * Outlier rejection filters single-frame TrackNet glitches before
//     velocity computation (useAdvancedAnalytics behavior).
//   * MAX_GAP_S (2.5s) guard prevents velocity from being built across
//     inter-rally idle periods.
//   * Stride subsampling ('auto'): when visible-shuttle coverage > 50% of
//     total frames, enable backend-style subsampling by keeping points at
//     least `stride` frames apart; below 50% coverage, use every point.
//
// Extra gates (opt-in, for auto-pause callers that need the stricter Layer
// B + C semantics from useShotSegments):
//   * minAccelMagPx:       reject candidates whose |v2 - v1| is below this
//                          (raw per-step velocity difference magnitude).
//   * wristProximityMeters: reject candidates where the nearest visible
//                           wrist is farther than this many court-meters
//                           from the shuttle. Requires `homography`.
// =============================================================================

import type {
  SkeletonFrame,
  FramePlayer,
  PoseClassificationResult,
} from '@/types/analysis'
import { applyHomography } from '@/utils/homography'

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------
export type DetectionMethod =
  | 'shuttle_trajectory'
  | 'pose_classification'
  | 'player_movement'
  | 'speed_peaks'

export interface ShotEvent {
  frame: number
  timestamp: number
  playerId: number
  shuttlePosition: { x: number; y: number } | null
  playerPosition: { x: number; y: number }
  detectionMethod: DetectionMethod
  shotType?: string | null
}

// -----------------------------------------------------------------------------
// Constants (exported for reuse)
// -----------------------------------------------------------------------------
const LEFT_WRIST_KP = 9
const RIGHT_WRIST_KP = 10

/** Minimum keypoint confidence for a wrist to be considered visible. */
const WRIST_CONFIDENCE_THRESHOLD = 0.3

/** Single-frame jump threshold (squared pixels) for TrackNet outlier rejection. */
const OUTLIER_DIST_SQ = 400 * 400

/**
 * Max time (seconds) between consecutive sampled shuttle positions before
 * the pair's velocity is treated as meaningless. Prevents inter-rally gaps
 * from producing phantom direction reversals.
 */
const MAX_GAP_S = 2.5

// -----------------------------------------------------------------------------
// Helpers
// -----------------------------------------------------------------------------

/** Find the player whose center is closest to (x, y). Returns null when none. */
function findClosestPlayer(
  players: FramePlayer[],
  x: number,
  y: number,
): FramePlayer | null {
  let closest: FramePlayer | null = null
  let closestDist = Infinity

  for (const player of players) {
    if (!player.center) continue
    const dx = player.center.x - x
    const dy = player.center.y - y
    const d = Math.sqrt(dx * dx + dy * dy)
    if (d < closestDist) {
      closestDist = d
      closest = player
    }
  }

  return closest ?? players[0] ?? null
}

/**
 * Smallest distance (in court meters) between the shuttle and any confident
 * wrist keypoint across the given players. Returns null when the homography
 * is missing, no wrist is visible, or a keypoint cannot be projected.
 */
function nearestWristMeters(
  players: FramePlayer[],
  shuttleX: number,
  shuttleY: number,
  H: number[][] | null,
): number | null {
  if (!H) return null

  const shuttleM = applyHomography(H, shuttleX, shuttleY)
  if (!shuttleM) return null

  let best: number | null = null
  for (const player of players) {
    const kpts = player.keypoints
    if (!kpts) continue
    for (const idx of [LEFT_WRIST_KP, RIGHT_WRIST_KP]) {
      const w = kpts[idx]
      if (!w?.x || !w?.y) continue
      if (w.confidence < WRIST_CONFIDENCE_THRESHOLD) continue
      const wM = applyHomography(H, w.x, w.y)
      if (!wM) continue
      const d = Math.hypot(shuttleM.x - wM.x, shuttleM.y - wM.y)
      if (best == null || d < best) best = d
    }
  }
  return best
}

// -----------------------------------------------------------------------------
// Shuttle trajectory detector
// -----------------------------------------------------------------------------

export interface DetectShuttleShotsOptions {
  /** Video frame rate (frames per second). Required for stride + gap math. */
  fps: number
  /** Forward homography (video pixels → court meters). Optional. */
  homography?: number[][] | null
  /** Minimum gap between consecutive shots in seconds. */
  minShotGapSec?: number
  /**
   * Minimum per-step displacement squared (px²). A candidate pair is rejected
   * if BOTH `|p1-p0|²` and `|p2-p1|²` are below this threshold (pure jitter).
   */
  minSpeedSq?: number
  /**
   * Angle threshold between consecutive velocity vectors.
   *   * >= 0 → require `dot < 0` (any reversal past 90°); value ignored.
   *   * < 0  → require `dot < cosAngleMax * sqrt(s1sq * s2sq)`.
   * Matches backend Python dot_threshold semantics.
   */
  cosAngleMax?: number
  /**
   * Optional. When set, reject candidates whose |v2 - v1| magnitude is below
   * this threshold. Useful for auto-pause to reject smooth gravity-arc flips.
   */
  minAccelMagPx?: number | null
  /**
   * Optional. When set AND `homography` is provided, reject candidates whose
   * nearest visible wrist is farther than this many court-meters from the
   * shuttle at the candidate frame.
   */
  wristProximityMeters?: number | null
  /** Filter single-frame shuttle outliers before velocity. Default: true. */
  rejectOutliers?: boolean
  /**
   * Stride subsampling control (seconds between kept samples).
   *   * number: always subsample with stride = max(3, round(fps * strideSec)).
   *   * 'auto': enable subsampling only when coverage > 50%, using the
   *            overhead-camera default stride of 0.3s.
   *   * null/undefined: no subsampling.
   */
  strideSec?: number | 'auto' | null
}

interface ShuttlePoint {
  frame: number
  timestamp: number
  x: number
  y: number
}

/**
 * Resolve overhead-camera defaults. Only fills values that are `undefined`.
 * The app only supports an overhead camera centered above the court — these
 * thresholds are tuned for that single viewpoint.
 */
function resolveShuttleDefaults(opts: DetectShuttleShotsOptions): {
  minShotGapSec: number
  minSpeedSq: number
  cosAngleMax: number
  autoStrideSec: number
} {
  return {
    minShotGapSec: opts.minShotGapSec ?? 0.6,
    minSpeedSq: opts.minSpeedSq ?? 225,
    cosAngleMax: opts.cosAngleMax ?? 0,
    autoStrideSec: 0.3,
  }
}

/**
 * Filter single-frame shuttle outliers. A position is dropped when its
 * squared distance to BOTH its predecessor and successor exceeds
 * OUTLIER_DIST_SQ (TrackNet glitches mis-locating the shuttle for one
 * frame). First and last points are always kept.
 */
function filterOutliers(points: ShuttlePoint[]): ShuttlePoint[] {
  if (points.length < 3) return points.slice()
  const out: ShuttlePoint[] = [points[0]!]
  for (let i = 1; i < points.length - 1; i++) {
    const prev = points[i - 1]!
    const curr = points[i]!
    const next = points[i + 1]!
    const dPrev = (curr.x - prev.x) ** 2 + (curr.y - prev.y) ** 2
    const dNext = (curr.x - next.x) ** 2 + (curr.y - next.y) ** 2
    if (dPrev > OUTLIER_DIST_SQ && dNext > OUTLIER_DIST_SQ) continue
    out.push(curr)
  }
  out.push(points[points.length - 1]!)
  return out
}

/**
 * Detect shots from shuttle trajectory direction reversals.
 *
 * Implements, in order:
 *   1. Collect visible shuttle positions.
 *   2. Optional outlier rejection (default on).
 *   3. Resolve overhead-camera defaults for undefined options.
 *   4. Optional stride subsampling ('auto' gates on >50% coverage).
 *   5. For each triple (p0, p1, p2):
 *        - skip when both per-step speeds are below `minSpeedSq`.
 *        - skip when MAX_GAP_S is exceeded between p0→p1 or p1→p2.
 *        - compute dot product; compare against threshold per `cosAngleMax`
 *          (backend-compatible formulation).
 *        - optional accel gate (|v2 - v1| < minAccelMagPx).
 *        - enforce min-shot-gap in frames.
 *        - lookup skeleton frame for p1; optional wrist-proximity gate.
 *        - assign to closest player and emit ShotEvent.
 */
export function detectShuttleShots(
  frames: SkeletonFrame[],
  opts: DetectShuttleShotsOptions,
): ShotEvent[] {
  const { fps, homography = null } = opts
  if (!fps || fps <= 0) return []

  const {
    minShotGapSec,
    minSpeedSq,
    cosAngleMax,
    autoStrideSec,
  } = resolveShuttleDefaults(opts)

  const minShotGapFrames = Math.max(3, Math.floor(fps * minShotGapSec))

  // ---------------------------------------------------------------------------
  // 1. Collect visible shuttle positions
  // ---------------------------------------------------------------------------
  const raw: ShuttlePoint[] = []
  for (const f of frames) {
    const s = f.shuttle_position
    if (s && s.x != null && s.y != null) {
      raw.push({ frame: f.frame, timestamp: f.timestamp, x: s.x, y: s.y })
    }
  }

  if (raw.length < 5) return []

  // ---------------------------------------------------------------------------
  // 2. Outlier rejection
  // ---------------------------------------------------------------------------
  const rejectOutliers = opts.rejectOutliers ?? true
  const cleaned = rejectOutliers ? filterOutliers(raw) : raw

  // ---------------------------------------------------------------------------
  // 3 + 4. Stride subsampling
  // ---------------------------------------------------------------------------
  let samples: ShuttlePoint[] = cleaned
  let strideFrames = 0

  const strideSec = opts.strideSec
  if (strideSec === 'auto') {
    const totalFrames = frames.length || 1
    const coverage = raw.length / totalFrames
    if (coverage > 0.5) {
      strideFrames = Math.max(3, Math.round(fps * autoStrideSec))
    }
  } else if (typeof strideSec === 'number') {
    strideFrames = Math.max(3, Math.round(fps * strideSec))
  }

  if (strideFrames > 0 && cleaned.length > 0) {
    const subsampled: ShuttlePoint[] = [cleaned[0]!]
    for (let i = 1; i < cleaned.length; i++) {
      if (cleaned[i]!.frame - subsampled[subsampled.length - 1]!.frame >= strideFrames) {
        subsampled.push(cleaned[i]!)
      }
    }
    samples = subsampled
  }

  if (samples.length < 3) return []

  // ---------------------------------------------------------------------------
  // 5. Candidate triples
  // ---------------------------------------------------------------------------
  const frameMap = new Map<number, SkeletonFrame>()
  for (const f of frames) frameMap.set(f.frame, f)

  const shots: ShotEvent[] = []
  let lastShotFrame = -Infinity

  for (let i = 2; i < samples.length; i++) {
    const p0 = samples[i - 2]!
    const p1 = samples[i - 1]!
    const p2 = samples[i]!

    // Max-time-gap guard: velocity is meaningless across long idle gaps.
    if ((p1.timestamp - p0.timestamp) > MAX_GAP_S || (p2.timestamp - p1.timestamp) > MAX_GAP_S) {
      continue
    }

    const vx1 = p1.x - p0.x
    const vy1 = p1.y - p0.y
    const vx2 = p2.x - p1.x
    const vy2 = p2.y - p1.y

    const speed1sq = vx1 * vx1 + vy1 * vy1
    const speed2sq = vx2 * vx2 + vy2 * vy2

    // Both near-stationary → jitter, not a hit.
    if (speed1sq < minSpeedSq && speed2sq < minSpeedSq) {
      continue
    }

    const dot = vx1 * vx2 + vy1 * vy2

    // Dot-threshold math — matches backend Python formulation.
    const threshold = cosAngleMax < 0
      ? cosAngleMax * Math.sqrt(speed1sq * speed2sq)
      : 0
    if (dot >= threshold) {
      continue
    }

    // Optional accel gate (auto-pause semantics).
    if (opts.minAccelMagPx != null) {
      const dvx = vx2 - vx1
      const dvy = vy2 - vy1
      const accelMag = Math.sqrt(dvx * dvx + dvy * dvy)
      if (accelMag < opts.minAccelMagPx) {
        continue
      }
    }

    // Min-shot-gap (frame-based, matches backend).
    if ((p1.frame - lastShotFrame) < minShotGapFrames) {
      continue
    }

    // Skeleton lookup for player assignment.
    const skFrame = frameMap.get(p1.frame)
    if (!skFrame || skFrame.players.length === 0) {
      continue
    }

    // Optional wrist-proximity gate.
    if (opts.wristProximityMeters != null && homography) {
      const wristDist = nearestWristMeters(skFrame.players, p1.x, p1.y, homography)
      if (wristDist != null && wristDist > opts.wristProximityMeters) {
        continue
      }
    }

    const closest = findClosestPlayer(skFrame.players, p1.x, p1.y)
    if (!closest) {
      continue
    }

    shots.push({
      frame: p1.frame,
      timestamp: p1.timestamp,
      playerId: closest.player_id,
      shuttlePosition: { x: p1.x, y: p1.y },
      playerPosition: closest.center ? { x: closest.center.x, y: closest.center.y } : { x: 0, y: 0 },
      detectionMethod: 'shuttle_trajectory',
    })
    lastShotFrame = p1.frame
  }

  return shots
}

// -----------------------------------------------------------------------------
// Pose classification fallback
// -----------------------------------------------------------------------------

export interface DetectPoseShotsOptions {
  /** Video frame rate (frames per second). */
  fps: number
  /** Set of class_name values considered hitting poses (e.g. 'smash'). */
  hittingClasses: Set<string>
  /** Minimum classification confidence. Default 0.5. */
  minConfidence?: number
  /** Inter-shot minimum gap across all players, in seconds. Default 0.5. */
  minShotGapSec?: number
  /** Same-player re-hit minimum gap in seconds. Default 0.8. */
  perPlayerGapSec?: number
}

/**
 * Detect shots from frame-level pose_classifications.
 *
 * For each frame with classifications:
 *   - Respect the overall `minShotGapSec` guard (frame-based).
 *   - For each classification, require class in `hittingClasses` and
 *     confidence >= `minConfidence`.
 *   - Match player: if the classification has a bbox AND the frame has
 *     multiple players, pick the player closest to the bbox center;
 *     otherwise default to `frame.players[0]`.
 *   - Respect a per-player gap (frame-based) to prevent double-counting a
 *     held pose.
 */
export function detectPoseShots(
  frames: SkeletonFrame[],
  opts: DetectPoseShotsOptions,
): ShotEvent[] {
  const { fps, hittingClasses } = opts
  if (!fps || fps <= 0) return []

  const minConfidence = opts.minConfidence ?? 0.5
  const minShotGapSec = opts.minShotGapSec ?? 0.5
  const perPlayerGapSec = opts.perPlayerGapSec ?? 0.8

  const minShotGapFrames = Math.max(1, Math.floor(fps * minShotGapSec))
  const perPlayerGapFrames = Math.max(minShotGapFrames, Math.floor(fps * perPlayerGapSec))

  const shots: ShotEvent[] = []
  const lastShotByPlayer = new Map<number, number>()
  let lastShotFrame = -Infinity

  for (const frame of frames) {
    const classifications = frame.pose_classifications
    if (!classifications || classifications.length === 0) continue
    if ((frame.frame - lastShotFrame) < minShotGapFrames) continue

    for (const cls of classifications as PoseClassificationResult[]) {
      if (!hittingClasses.has(cls.class_name)) continue
      if (cls.confidence < minConfidence) continue

      let matchedPlayer: FramePlayer | null = frame.players[0] ?? null

      if (cls.bbox && frame.players.length > 1) {
        const cx = cls.bbox.x + cls.bbox.width / 2
        const cy = cls.bbox.y + cls.bbox.height / 2
        matchedPlayer = findClosestPlayer(frame.players, cx, cy) ?? matchedPlayer
      }

      if (!matchedPlayer) continue

      const lastForPlayer = lastShotByPlayer.get(matchedPlayer.player_id) ?? -Infinity
      if ((frame.frame - lastForPlayer) < perPlayerGapFrames) continue

      const shuttle = frame.shuttle_position
      shots.push({
        frame: frame.frame,
        timestamp: frame.timestamp,
        playerId: matchedPlayer.player_id,
        shuttlePosition: shuttle && shuttle.x != null && shuttle.y != null
          ? { x: shuttle.x, y: shuttle.y }
          : null,
        playerPosition: matchedPlayer.center
          ? { x: matchedPlayer.center.x, y: matchedPlayer.center.y }
          : { x: 0, y: 0 },
        detectionMethod: 'pose_classification',
        shotType: cls.class_name,
      })
      lastShotFrame = frame.frame
      lastShotByPlayer.set(matchedPlayer.player_id, frame.frame)
      // Only one shot per frame.
      break
    }
  }

  return shots
}

// -----------------------------------------------------------------------------
// Merge
// -----------------------------------------------------------------------------

/**
 * Merge shot events from multiple detectors, preferring shuttle-detected
 * shots over pose-only ones when they collide within `minGapFrames`.
 *
 * Mirrors useAdvancedAnalytics.ts:mergeShots:
 *   - Sort by frame.
 *   - Walk in order; if the current shot is within minGapFrames of the last
 *     accepted one, keep whichever has a non-null `shuttlePosition` (shuttle
 *     wins); otherwise drop the current shot.
 */
export function mergeShots(shots: ShotEvent[], minGapFrames: number): ShotEvent[] {
  const sorted = shots.slice().sort((a, b) => a.frame - b.frame)
  const merged: ShotEvent[] = []
  for (const shot of sorted) {
    const last = merged[merged.length - 1]
    if (last && (shot.frame - last.frame) < minGapFrames) {
      if (last.shuttlePosition == null && shot.shuttlePosition != null) {
        merged[merged.length - 1] = shot
      }
      continue
    }
    merged.push(shot)
  }
  return merged
}

File: backend/shot_detection.py
"""
Shot detection — Python port of src/utils/shotDetection.ts.

Mirrors the client-side detector that powers in-browser rally segmentation
in useAdvancedAnalytics.ts. Used by rally_detection_shot_gap.py to feed the
shot-gap rally detector that runs alongside the gradient-based detector
(rally_detection.py) in the Modal worker.

Algorithmic parity with the TS source:
  * Outlier rejection on visible shuttle positions (single-frame TrackNet
    glitches dropped before velocity computation).
  * Stride='auto': enable backend-style subsampling (>50% coverage) with
    autoStrideSec=0.3.
  * Dot-threshold math: cosAngleMax<0 ⇒ threshold = cosAngleMax * sqrt(s1²s2²);
    otherwise threshold = 0. Trigger when dot < threshold.
  * MAX_GAP_S=2.5 guard against velocity built across inter-rally idle.
  * minSpeedSq=225 (15 px) — both per-step speeds below ⇒ jitter, skip.
  * minShotGapFrames = max(3, floor(fps * 0.6)).
  * Pose-classification fallback (detect_pose_shots) when shuttle data is
    too sparse, then mergeShots() preferring shuttle-detected shots.
  * Wrist-proximity gate is OFF for the rally use case (TS sets
    wristProximityMeters=null), so the Python port omits homography wholly.
    The TS rally caller also passes minAccelMagPx=null, so the accel gate
    is omitted too. (These options exist in the TS source for the
    auto-pause caller, which is not ported here.)

Frame field mapping (TS SkeletonFrame → Python skeleton_frame dict):
  TS f.frame              → py f["frame"]
  TS f.timestamp          → py f["timestamp"]
  TS f.shuttle_position   → py f.get("shuttle_position")  (or None)
  TS f.players[i].player_id → py f["players"][i]["player_id"]
  TS f.players[i].center  → py f["players"][i].get("center")
  TS f.players[i].keypoints[idx]  → py f["players"][i]["keypoints"][idx]
       Note: TS uses dense index lookup (kpts[9] = left wrist). Python
       skeleton_frames also store keypoints as an ordered list of
       {name, x, y, confidence} dicts, indexed by COCO keypoint id, so
       direct integer indexing works. We still bounds-check.
  TS f.pose_classifications → py f.get("pose_classifications") (rarely set
       in the Python pipeline, but we honor it for parity).
"""

from __future__ import annotations

import math
from typing import Any, Dict, List, Optional, Tuple


# -----------------------------------------------------------------------------
# Constants — keep aligned with src/utils/shotDetection.ts
# -----------------------------------------------------------------------------
LEFT_WRIST_KP = 9
RIGHT_WRIST_KP = 10

# Min keypoint confidence for a wrist to be considered visible.
WRIST_CONFIDENCE_THRESHOLD = 0.3

# Single-frame jump threshold (squared pixels) for outlier rejection.
OUTLIER_DIST_SQ = 400 * 400

# Max time between sampled shuttle positions before velocity is meaningless.
MAX_GAP_S = 2.5

# High-confidence hitting poses — only poses that unambiguously indicate a
# shot. Mirrors HITTING_POSES in useAdvancedAnalytics.ts.
HITTING_POSES = {"smash", "overhead", "serving", "serve", "offense"}


# -----------------------------------------------------------------------------
# Helpers
# -----------------------------------------------------------------------------
def _find_closest_player(
    players: List[Dict[str, Any]], x: float, y: float
) -> Optional[Dict[str, Any]]:
    """Find the player whose center is closest to (x, y). Returns None when
    none have a center; falls back to players[0] when all centers are
    missing — same fallback as the TS findClosestPlayer."""
    closest: Optional[Dict[str, Any]] = None
    closest_dist = math.inf

    for player in players:
        center = player.get("center")
        if not center:
            continue
        cx = center.get("x")
        cy = center.get("y")
        if cx is None or cy is None:
            continue
        dx = cx - x
        dy = cy - y
        d = math.sqrt(dx * dx + dy * dy)
        if d < closest_dist:
            closest_dist = d
            closest = player

    if closest is not None:
        return closest
    return players[0] if players else None


def _filter_outliers(points: List[Tuple[int, float, float, float]]) -> List[
    Tuple[int, float, float, float]
]:
    """Drop a position when its squared distance to BOTH neighbors exceeds
    OUTLIER_DIST_SQ. First and last are always kept.

    Each point is (frame, timestamp, x, y).
    """
    if len(points) < 3:
        return list(points)
    out: List[Tuple[int, float, float, float]] = [points[0]]
    for i in range(1, len(points) - 1):
        prev = points[i - 1]
        curr = points[i]
        nxt = points[i + 1]
        d_prev = (curr[2] - prev[2]) ** 2 + (curr[3] - prev[3]) ** 2
        d_next = (curr[2] - nxt[2]) ** 2 + (curr[3] - nxt[3]) ** 2
        if d_prev > OUTLIER_DIST_SQ and d_next > OUTLIER_DIST_SQ:
            continue
        out.append(curr)
    out.append(points[-1])
    return out


# -----------------------------------------------------------------------------
# Shuttle trajectory detector
# -----------------------------------------------------------------------------
def detect_shuttle_shots(
    frames: List[Dict[str, Any]],
    fps: float,
    min_shot_gap_sec: float = 0.6,
    min_speed_sq: float = 225.0,
    cos_angle_max: float = 0.0,
    reject_outliers: bool = True,
    stride_sec: Any = "auto",
    auto_stride_sec: float = 0.3,
) -> List[Dict[str, Any]]:
    """Direct port of detectShuttleShots(frames, opts).

    Returns a list of shot-event dicts. The TS ShotEvent fields are mapped to
    snake_case here for Python convention, but the rally-grouping code only
    reads `frame` and `timestamp`, so naming downstream is not load-bearing.
    """
    if not fps or fps <= 0:
        return []

    min_shot_gap_frames = max(3, int(fps * min_shot_gap_sec))

    # 1. Collect visible shuttle positions
    raw: List[Tuple[int, float, float, float]] = []
    for f in frames:
        s = f.get("shuttle_position")
        if s and s.get("x") is not None and s.get("y") is not None:
            raw.append((f["frame"], f["timestamp"], s["x"], s["y"]))

    if len(raw) < 5:
        return []

    # 2. Outlier rejection
    cleaned = _filter_outliers(raw) if reject_outliers else raw

    # 3 + 4. Stride subsampling
    samples: List[Tuple[int, float, float, float]] = cleaned
    stride_frames = 0

    if stride_sec == "auto":
        total_frames = len(frames) or 1
        coverage = len(raw) / total_frames
        if coverage > 0.5:
            stride_frames = max(3, round(fps * auto_stride_sec))
    elif isinstance(stride_sec, (int, float)) and stride_sec is not None:
        stride_frames = max(3, round(fps * stride_sec))

    if stride_frames > 0 and cleaned:
        subsampled: List[Tuple[int, float, float, float]] = [cleaned[0]]
        for pt in cleaned[1:]:
            if pt[0] - subsampled[-1][0] >= stride_frames:
                subsampled.append(pt)
        samples = subsampled

    if len(samples) < 3:
        return []

    # 5. Candidate triples
    frame_map: Dict[int, Dict[str, Any]] = {f["frame"]: f for f in frames}

    shots: List[Dict[str, Any]] = []
    last_shot_frame = -math.inf

    for i in range(2, len(samples)):
        f0, t0, x0, y0 = samples[i - 2]
        f1, t1, x1, y1 = samples[i - 1]
        f2, t2, x2, y2 = samples[i]

        # Max-time-gap guard.
        if (t1 - t0) > MAX_GAP_S or (t2 - t1) > MAX_GAP_S:
            continue

        vx1 = x1 - x0
        vy1 = y1 - y0
        vx2 = x2 - x1
        vy2 = y2 - y1

        speed1_sq = vx1 * vx1 + vy1 * vy1
        speed2_sq = vx2 * vx2 + vy2 * vy2

        if speed1_sq < min_speed_sq and speed2_sq < min_speed_sq:
            continue

        dot = vx1 * vx2 + vy1 * vy2

        if cos_angle_max < 0:
            threshold = cos_angle_max * math.sqrt(speed1_sq * speed2_sq)
        else:
            threshold = 0.0
        if dot >= threshold:
            continue

        # (Accel gate omitted — rally caller passes minAccelMagPx=null.)

        if (f1 - last_shot_frame) < min_shot_gap_frames:
            continue

        sk = frame_map.get(f1)
        if not sk:
            continue
        players = sk.get("players") or []
        if not players:
            continue

        # (Wrist-proximity gate omitted — rally caller passes
        # wristProximityMeters=null and homography is not available
        # backend-side at this stage anyway.)

        closest = _find_closest_player(players, x1, y1)
        if not closest:
            continue

        center = closest.get("center") or {"x": 0.0, "y": 0.0}

        shots.append({
            "frame": int(f1),
            "timestamp": float(t1),
            "player_id": int(closest.get("player_id", 0)),
            "shot_type": "unknown",
            "shuttle_position": {"x": float(x1), "y": float(y1)},
            "player_position": {
                "x": float(center.get("x", 0.0)),
                "y": float(center.get("y", 0.0)),
            },
            "detection_method": "shuttle_trajectory",
        })
        last_shot_frame = f1

    return shots


# -----------------------------------------------------------------------------
# Pose classification fallback
# -----------------------------------------------------------------------------
def detect_pose_shots(
    frames: List[Dict[str, Any]],
    fps: float,
    hitting_classes: Optional[set] = None,
    min_confidence: float = 0.65,
    min_shot_gap_sec: float = 0.6,
    per_player_gap_sec: float = 0.8,
) -> List[Dict[str, Any]]:
    """Direct port of detectPoseShots. Reads frame.pose_classifications.

    Note: the Python skeleton_frames pipeline currently does NOT populate
    pose_classifications (per the doc-comment in useAdvancedAnalytics.ts:
    "NEVER POPULATED"), so this fallback will typically yield no shots.
    Kept for parity in case the pipeline starts emitting them.
    """
    if not fps or fps <= 0:
        return []
    if hitting_classes is None:
        hitting_classes = HITTING_POSES

    min_shot_gap_frames = max(1, int(fps * min_shot_gap_sec))
    per_player_gap_frames = max(min_shot_gap_frames, int(fps * per_player_gap_sec))

    shots: List[Dict[str, Any]] = []
    last_shot_by_player: Dict[int, int] = {}
    last_shot_frame = -10**9

    for frame in frames:
        classifications = frame.get("pose_classifications")
        if not classifications:
            continue
        if (frame["frame"] - last_shot_frame) < min_shot_gap_frames:
            continue

        players = frame.get("players") or []

        for cls in classifications:
            cname = cls.get("class_name")
            if cname not in hitting_classes:
                continue
            if cls.get("confidence", 0.0) < min_confidence:
                continue

            matched: Optional[Dict[str, Any]] = players[0] if players else None

            bbox = cls.get("bbox")
            if bbox and len(players) > 1:
                cx = bbox["x"] + bbox["width"] / 2
                cy = bbox["y"] + bbox["height"] / 2
                matched = _find_closest_player(players, cx, cy) or matched

            if not matched:
                continue

            pid = int(matched.get("player_id", 0))
            last_for_player = last_shot_by_player.get(pid, -10**9)
            if (frame["frame"] - last_for_player) < per_player_gap_frames:
                continue

            shuttle = frame.get("shuttle_position")
            shuttle_pos = None
            if shuttle and shuttle.get("x") is not None and shuttle.get("y") is not None:
                shuttle_pos = {"x": float(shuttle["x"]), "y": float(shuttle["y"])}

            center = matched.get("center") or {"x": 0.0, "y": 0.0}

            shots.append({
                "frame": int(frame["frame"]),
                "timestamp": float(frame["timestamp"]),
                "player_id": pid,
                "shot_type": cname,
                "shuttle_position": shuttle_pos,
                "player_position": {
                    "x": float(center.get("x", 0.0)),
                    "y": float(center.get("y", 0.0)),
                },
                "detection_method": "pose_classification",
            })
            last_shot_frame = frame["frame"]
            last_shot_by_player[pid] = frame["frame"]
            # Only one shot per frame.
            break

    return shots


# -----------------------------------------------------------------------------
# Merge
# -----------------------------------------------------------------------------
def merge_shots(
    shots: List[Dict[str, Any]], min_gap_frames: int
) -> List[Dict[str, Any]]:
    """Direct port of mergeShots. Sort by frame; if the current shot is
    within min_gap_frames of the last accepted, prefer whichever has a
    non-null shuttle_position."""
    sorted_shots = sorted(shots, key=lambda s: s["frame"])
    merged: List[Dict[str, Any]] = []
    for shot in sorted_shots:
        if merged:
            last = merged[-1]
            if (shot["frame"] - last["frame"]) < min_gap_frames:
                if last.get("shuttle_position") is None and shot.get("shuttle_position") is not None:
                    merged[-1] = shot
                continue
        merged.append(shot)
    return merged


# -----------------------------------------------------------------------------
# Top-level entry point — mirrors useAdvancedAnalytics.detectAllShots
# -----------------------------------------------------------------------------
def detect_all_shots(frames: List[Dict[str, Any]], fps: float) -> List[Dict[str, Any]]:
    """Mirrors useAdvancedAnalytics.detectAllShots(frames, fps).

    Strategy:
      1. Run shuttle-trajectory detector (rally use-case options: wrist gate
         off, accel gate off, outlier rejection on, stride='auto').
      2. If we got >= 4 shuttle shots, return those (and tag shot_type
         'unknown' to match the TS toRallyShot output).
      3. Otherwise add the pose-classification fallback and merge,
         preferring shuttle-detected shots.

    Each returned shot dict has: frame, timestamp, player_id, shot_type,
    shuttle_position, player_position, detection_method.
    """
    shuttle_shots = detect_shuttle_shots(
        frames,
        fps=fps,
        reject_outliers=True,
        stride_sec="auto",
    )

    if len(shuttle_shots) >= 4:
        return shuttle_shots

    pose_shots = detect_pose_shots(
        frames,
        fps=fps,
        hitting_classes=HITTING_POSES,
        min_confidence=0.65,
        min_shot_gap_sec=0.6,
        per_player_gap_sec=0.8,
    )

    if not pose_shots:
        return shuttle_shots

    min_gap_frames = max(3, int(fps * 0.6))
    return merge_shots(shuttle_shots + pose_shots, min_gap_frames)
